src/exact.jl

```diff
@@ -0,0 +1,319 @@
+a sparse matrix.
+
+See also: [`ot_cost`](@ref), [`emd`](@ref)
+"""
+function ot_plan(_, μ::DiscreteNonParametric, ν::DiscreteNonParametric)
+    # unpack the probabilities of the two distributions
+    μprobs = probs(μ)
+    νprobs = probs(ν)
+
+    # create the iterator
+    # note: support of `DiscreteNonParametric` is sorted
+    iter = Discrete1DOTIterator(μprobs, νprobs)
+
+    # create arrays for the indices of the two histograms and the optimal flow between the
+    # corresponding points
+    I = Int[]
+    J = Int[]
+    W = Vector{Base.promote_eltype(μprobs, νprobs)}(undef, 0)
+
+    # pre-allocate some space to speed up `push!` in the iteration below
+    minsize = length(μprobs) + length(νprobs)
+    sizehint!(I, minsize)
+    sizehint!(J, minsize)
+    sizehint!(W, minsize)
+
+    # compute the sparse optimal transport plan
+    for (i, j, w) in iter
+        push!(I, i)
+        push!(J, j)
+        push!(W, w)
+    end
+    γ = sparse(I, J, W, length(μprobs), length(νprobs))
+
+    return γ
+end
+
+"""
+    ot_cost(
+        c, μ::DiscreteNonParametric, ν::DiscreteNonParametric; plan=nothing
+    )
+
+Compute the optimal transport cost for the Monge-Kantorovich problem with discrete
+univariate distributions `μ` and `ν` as source and target marginals and cost function `c`
+of the form ``c(x, y) = h(|x - y|)`` where ``h`` is a convex function.
+
+In this setting, the optimal transport cost can be computed analytically.
+
+A pre-computed optimal transport `plan` may be provided.
+
+See also: [`ot_plan`](@ref), [`emd2`](@ref)
+"""
+function ot_cost(c, μ::DiscreteNonParametric, ν::DiscreteNonParametric; plan=nothing)
+    return _ot_cost(c, μ, ν, plan)
+end
+
+# compute cost from scratch if no plan is provided
+function _ot_cost(c, μ::DiscreteNonParametric, ν::DiscreteNonParametric, ::Nothing)
+    # unpack the probabilities of the two distributions
+    μprobs = probs(μ)
+    νprobs = probs(ν)
+
+    # create the iterator
+    # note: support of `DiscreteNonParametric` is sorted
+    iter = Discrete1DOTIterator(μprobs, νprobs)
+
+    # compute the cost
+    μsupport = support(μ)
+    νsupport = support(ν)
+    cost = sum(w * c(μsupport[i], νsupport[j]) for (i, j, w) in iter)
+
+    return cost
+end
+
+# if a sparse plan is provided, we just iterate through the non-zero entries
+function _ot_cost(
+    c, μ::DiscreteNonParametric, ν::DiscreteNonParametric, plan::SparseMatrixCSC
+)
+    # extract non-zero flows
+    I, J, W = findnz(plan)
+
+    # compute the cost
+    μsupport = support(μ)
+    νsupport = support(ν)
+    cost = sum(w * c(μsupport[i], νsupport[j]) for (i, j, w) in zip(I, J, W))
+
+    return cost
+end
+
+# fallback: compute cost matrix (probably often faster to compute cost from scratch)
+function _ot_cost(c, μ::DiscreteNonParametric, ν::DiscreteNonParametric, plan)
+    return dot(plan, StatsBase.pairwise(c, support(μ), support(ν)))
+end
```

Comment:
Did your benchmark show that this `function _ot_cost( c, μ::DiscreteNonParametric, ν::DiscreteNonParametric, plan::SparseMatrixCSC)` was more efficient than the fallback using ` dot(plan, StatsBase.pairwise(c, support(μ), support(ν)))`? Because when I was testing my code, I kept getting better results for the latter, even for sparse matrices.

Reply:
No, but I'll perform some benchmarks. I assume it depends on the size of mu and nu - the benefit of the sparse matrix and avoiding all pairwise computations should be higher if one deals with large matrices. Additionally, probably also the sparse approach performs better with non-standard `c` since `pairwise` is optimized for `Euclidean` and `SqEuclidean`.

Reply:
I did some benchmarks: https://gist.github.com/devmotion/94ff735b13bab55ed051e8db8c906380

Short summary: For 10-dimensional mu and nu, the `pairwise` evaluation is slightly faster than the sparse iteration with `SqEuclidean` but already 10% slower with `Euclidean`. For 15-dimensional mu and nu, also with `SqEuclidean` the iteration is approx 30% faster than the `pairwise` evaluation. Of course, the difference increases with increasing dimension: For 200-dimensional mu and nu the sparse iteration is > 90% faster for both `Euclidean` and `SqEuclidean`.